### PR TITLE
OpenStack: get rid of swift temp urls

### PR DESCRIPTION
**Pull request by Fedosin**
Since not all the configurations of OSP support Swift temp urls, we have to find another way that will work on all supported clouds.

To create a unified solution we can use the ACL mechanism to allow public access to the ignition file, but deny access to the list of container files. From a technical point of view, it looks like this: ".r:*", but no ".rlistings".
This allows anybody to download bootstrap ignition file. However, to do this, the exact name of the object must be known since users cannot list the objects in the container.

The last step is to randomize the file name so that it cannot be matched - now it is hardcoded to "bootstrap.ign"

The lifetime of such a file will be short, about 10 minutes necessary for bootstrapping. Further this file will be automatically removed by Terraform at the moment of bootstrap machine destruction.

This way we can achieve the same effect as temp url, which will work on all types of storage.

### CODE REVIEW

**Comment by Fedosin** [comment]
/hold

we need to agree that we need this patch at 4.2

File: pkg/tfvars/openstack/openstack.go [hunk 1]
```diff
@@ -35,7 +45,72 @@ func TFVars(masterConfig *v1alpha1.OpenstackProviderSpec, cloud string, region s
...
+	serviceCatalog, err := getServiceCatalog(cloud)
+	if err != nil {
+		return "", err
+	}
+
+	for _, svc := range serviceCatalog.Entries {
+		if svc.Type == "object-store" {
+			for _, e := range svc.Endpoints {
+				if e.Interface == "public" {
+					swiftPublicURL = e.URL
```
Comment (celebdor) [inline pkg/tfvars/openstack/openstack.go]: shouldn't you break here? why going over the rest?

Comment (Fedosin) [reply pkg/tfvars/openstack/openstack.go]: yep, you're right

**Comment by MartijnStraatman** [comment]
I pulled this PR locally and tried to use it. First run was sucessfull. Second fails with this error:

```
time="2019-09-05T13:48:54-04:00" level=error msg="Error: Error creating OpenStack container object: parse https://<url>/swift/v1/<token>/os-test-tgb8j/aDAT#dtzs2%[Vm%G: invalid URL escape \"%[V\""
time="2019-09-05T13:48:54-04:00" level=error
time="2019-09-05T13:48:54-04:00" level=error msg="  on ../../../tmp/openshift-install-026823105/bootstrap/main.tf line 54, in resource \"openstack_objectstorage_object_v1\" \"ignition\":"
time="2019-09-05T13:48:54-04:00" level=error msg="  54: resource \"openstack_objectstorage_object_v1\" \"ignition\" {"
time="2019-09-05T13:48:54-04:00" level=error
time="2019-09-05T13:48:54-04:00" level=error
time="2019-09-05T13:48:54-04:00" level=fatal msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed to apply using Terraform"

**Comment by Fedosin** [comment]
> I pulled this PR locally and tried to use it. First run was sucessfull. Second fails with this error:
> 
> ```
> time="2019-09-05T13:48:54-04:00" level=error msg="Error: Error creating OpenStack container object: parse https://<url>/swift/v1/<token>/os-test-tgb8j/aDAT#dtzs2%[Vm%G: invalid URL escape \"%[V\""
> time="2019-09-05T13:48:54-04:00" level=error
> time="2019-09-05T13:48:54-04:00" level=error msg="  on ../../../tmp/openshift-install-026823105/bootstrap/main.tf line 54, in resource \"openstack_objectstorage_object_v1\" \"ignition\":"
> time="2019-09-05T13:48:54-04:00" level=error msg="  54: resource \"openstack_objectstorage_object_v1\" \"ignition\" {"
> time="2019-09-05T13:48:54-04:00" level=error
> time="2019-09-05T13:48:54-04:00" level=error
> time="2019-09-05T13:48:54-04:00" level=fatal msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed to apply using Terraform"
> ```

yeah, we need to use alphanumeric characters for the object name. Thank you for finding this!

**Comment by sdodson** [comment]
@celebdor PTAL or delegate to someone else for review

**Comment by tomassedovic** [comment]
/assign

I'm testing and reviewing it as we speak.

**Comment by tomassedovic** [comment]
Okay, I like it. I would really prefer if we kept using Swift temp urls since their behaviour and security profile are well understood, but not supporting OSP 13 + Ceph's RADOS Gateway would cut off a significant portion of the production users.

That said, I am no security expert and as such I am a little nervous about this change -- especially this late in the game.

To the best of my abilities, I've verified that:
* The Swift object name appears different and unpredictable on every deployment
* [The Terraform `random_string` provider uses a cryptographically secure random generator](https://www.terraform.io/docs/providers/random/r/string.html)
* An unauthorised user cannot list the container's objects (and therefore find the name and access the sensitive Ignition contents)
* The object is deleted after the bootstrapping phase so it doesn't linger indefinitely

There are some things I can see to make it even more solid:

1. Use the `random_password` resource instead of `random_string`. They should behave identically, but the password one is not echoed in any of the Terraform logs which means it won't leak in the CI (https://www.terraform.io/docs/providers/random/r/password.html)
2. Consider setting a timeout (the `delete_after` property) on the bootstrap ignition object (https://www.terraform.io/docs/providers/openstack/r/objectstorage_object_v1.html#delete_after). This will make the ignition inaccessible even if the bootstrap resources never get deleted (e.g. in the case of a failure). Probably not a huge issue, just tightening up.

I am tentatively in favour of this but I'd like to hold off from `lgtm` before more people have their say.

//cc @mandre (what do you think about the risks present here) and @racedo (what do you think about the impact of not shipping this feature in 4.2)

**Comment by Fedosin** [comment]
+1. thanks for testing this @tomassedovic !

**Comment by mandre** [comment]
/hold

Thanks for the great thorough review Tomas - I agree with everything you wrote.
I carefully reviewed the code myself, and this has dissipated my fears around security. I do not have any particular concerns merging this code.

After implementing your 2 suggesting, I think we'll have emulated the swift temp-url feature, without making it less secure.

Holding on for now to give @racedo a chance to chime in.

**Comment by Fedosin** [comment]
done :)

**Comment by Fedosin** [comment]
I had to bump terraform-provider-random because it did not support random_password

File: docs/user/openstack/README.md [hunk 2]
```diff
@@ -81,13 +81,6 @@ Swift is required for installation as the user-data provided by OpenStack Metada
 
 Swift is also used as a backend for the OpenShift image registry, but at the time of installation only an empty container is created without loading any data. Later on, for the system to work properly, you need to have enough free space to store the container images.
 
-The user must have `swiftoperator` permissions and `temp-url` support must be enabled. As an OpenStack administrator:
-
-```sh
-openstack role add --user <user> --project <project> swiftoperator
```
Comment (mandre) [inline docs/user/openstack/README.md]: AFAIK we still need the `swiftoperator` role to be able to create containers.

Comment (Fedosin) [reply docs/user/openstack/README.md]: yep

**Comment by mandre** [comment]
/hold

We decided to wait for first z-stream release of 4.2.

File: docs/user/openstack/README.md (hunk 2 above)
Comment (augustrh) [reply docs/user/openstack/README.md]: I don't think it exists for director-deployed RGW installs. In that case the user needs Member

File: pkg/tfvars/openstack/openstack.go [hunk 3]
```diff
@@ -62,3 +73,69 @@ func TFVars(masterConfig *v1alpha1.OpenstackProviderSpec, cloud string, external
...
+	}
+
+	authOptions := tokens.AuthOptions{
+		Scope:      scope,
+		Username:   cloudConfig.AuthInfo.Username,
+		Password:   cloudConfig.AuthInfo.Password,
+		DomainName: domainName,
+	}
+
+	serviceCatalog, err := tokens.Create(conn, &authOptions).ExtractServiceCatalog()
```
Comment (abhinavdahiya) [inline pkg/tfvars/openstack/openstack.go]: Hmm, i'm not sure what this does ? but does it create some resource that needs to be cleaned up on destroy?

File: pkg/tfvars/openstack/openstack.go [hunk 4]
```diff
@@ -62,3 +73,69 @@ func TFVars(masterConfig *v1alpha1.OpenstackProviderSpec, cloud string, external
...
+	}
+
+	if swiftPublicURL == "" {
+		return "", fmt.Errorf("cannot retrieve Swift URL from the service catalog")
+	}
+
+	return swiftPublicURL, nil
+}
+
+func getServiceCatalog(cloud string) (*tokens.ServiceCatalog, error) {
```
Comment (abhinavdahiya) [inline pkg/tfvars/openstack/openstack.go]: would love to see some comments on the function regarding what it does.

File: pkg/tfvars/openstack/openstack.go [hunk 5]
```diff
@@ -62,3 +73,69 @@ func TFVars(masterConfig *v1alpha1.OpenstackProviderSpec, cloud string, external
 
 	return json.MarshalIndent(cfg, "", "  ")
 }
+
+func getSwiftPublicURL(cloud string) (string, error) {
```
Comment (abhinavdahiya) [inline pkg/tfvars/openstack/openstack.go]: same as https://github.com/openshift/installer/pull/2311/commits/5c3fbf5fe0f0c15fa0889452a25263382ef4049b#r337229011

File: pkg/tfvars/openstack/openstack.go (hunk 3 above)
Comment (Fedosin) [reply pkg/tfvars/openstack/openstack.go]: Here we want to obtain Swift public endpoint... By design this should be done by using https://www.terraform.io/docs/providers/openstack/d/identity_endpoint_v3.html but OpenStack default policies forbid to use this API for regular users.
On the other hand when you authenticate in OpenStack (i.e. get a token) it includes the whole service catalog in the output json. So we are able to parse it and get the endpoint from there https://docs.openstack.org/api-ref/identity/v3/?expanded=token-authentication-with-scoped-authorization-detail#token-authentication-with-scoped-authorization
In short the algorithm is:
1. Authenticate in OpenStack: `tokens.Create(..)`
2. Parse the token and extract the service catalog: `ExtractServiceCatalog()`
3. Iterate through the catalog and find `public` endpoint for `object-store`.

Unfortunately this feature is not supported by Terraform so I had to implement it here. No resources are created we just get endpoint data from OpenStack.

Btw, we do the same thing for Octavia service: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/openstack/openstack.go#L141-L152

File: pkg/tfvars/openstack/openstack.go (hunk 4 above)
Comment (Fedosin) [reply pkg/tfvars/openstack/openstack.go]: done

File: pkg/tfvars/openstack/openstack.go (hunk 5 above)
Comment (Fedosin) [reply pkg/tfvars/openstack/openstack.go]: done

**Comment by abhinavdahiya** [comment]
8d04568695ff880e180b8719f2bde8347eede4e5 LGTM

/approve

while the rest i'll leave upto the openstack team to review

**Comment by MartijnStraatman** [comment]
will this PR be available in 4.2? If we install 4.2 by compiling the 4.2 release this feature seems to be missing.
